tracee-ebpf/tracee/events.go

```diff
@@ -0,0 +1,184 @@
+package tracee
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+	"unsafe"
+
+	"github.com/aquasecurity/tracee/tracee-ebpf/external"
+)
+
+type RawEvent struct {
+	Ctx      context
+	Args     map[string]interface{}
+	ArgMetas []external.ArgMeta
+}
+
+// context struct contains common metadata that is collected for all types of events
+// it is used to unmarshal binary data and therefore should match (bit by bit) to the `context_t` struct in the ebpf code.
+// NOTE: Integers want to be aligned in memory, so if changing the format of this struct
+// keep the 1-byte 'Argnum' as the final parameter before the padding (if padding is needed).
+type context struct {
+	Ts       uint64
+	Pid      uint32
+	Tid      uint32
+	Ppid     uint32
+	HostPid  uint32
+	HostTid  uint32
+	HostPpid uint32
+	Uid      uint32
+	MntID    uint32
+	PidID    uint32
+	Comm     [16]byte
+	UtsName  [16]byte
+	ContID   [16]byte
+	EventID  int32
+	Retval   int64
+	StackID  uint32
+	Argnum   uint8
+	_        [3]byte //padding
+}
+
+func (t *Tracee) processEvents(done <-chan struct{}) error {
+	for dataRaw := range t.eventsChannel {
+		dataBuff := bytes.NewBuffer(dataRaw)
+		var ctx context
+		err := binary.Read(dataBuff, binary.LittleEndian, &ctx)
+		if err != nil {
+			t.handleError(err)
+			continue
+		}
+
+		rawEvent := RawEvent{
+			Ctx:      ctx,
+			Args:     make(map[string]interface{}, ctx.Argnum),
+			ArgMetas: make([]external.ArgMeta, ctx.Argnum),
+		}
+
+		params := EventsIDToParams[ctx.EventID]
+		if params == nil {
+			t.handleError(fmt.Errorf("failed to get parameters of event %d", ctx.EventID))
+			continue
+		}
+
+		for i := 0; i < int(ctx.Argnum); i++ {
+			argMeta, argVal, err := readArgFromBuff(dataBuff, params)
+			if err != nil {
+				t.handleError(fmt.Errorf("failed to read argument %d of event %d: %v", i, ctx.EventID, err))
+				continue
+			}
+
+			rawEvent.Args[argMeta.Name] = argVal
+			rawEvent.ArgMetas[i] = argMeta
+		}
+
+		if !t.shouldProcessEvent(rawEvent) {
+			continue
+		}
+		err = t.processEvent(&rawEvent.Ctx, rawEvent.Args)
```

Comment:
If you keep the type creation (which I don't think is needed)... can't you pass &rawEvent only here ?

Reply:
What do you mean "keep the type creation"? Creating RawEvent? If so, then let me explain why it is created.
`RawEvent` (which shouldn't be with capital 'R' anymore - I'll fix it) is used to avoid iteration over the arguments.
The reason for that is that `RawEvent` contains an arguments map, which maps an argument name to its value. If we only created `external.Argument` slice (which is ordered by the argument number, unlike the arguments map), we had to iterate over it to get a specific argument value. So creating this `RawEvent` struct avoid this extra iteration, and simplifies the code (for example, in `prepareArgs()`).

> can't you pass &rawEvent only here

Yes, let me take care of that.